analyses/focal-cn-file-preparation/05-define-most-focal-cn-units.Rmd

```diff
@@ -0,0 +1,394 @@
+
+```{r}
+# Read in the file with consensus CN status data and the UCSC cytoband data --
+# generated in `03-add-cytoband-status-consensus.Rmd`
+consensus_seg_cytoband_status_df <-
+  read_tsv(file.path("results", "consensus_seg_with_ucsc_cytoband_status.tsv.gz")) %>%
+  # Need this to not have `chr`
+  mutate(
+    chr = gsub("chr", "", chr),
+    cytoband = paste0(chr, cytoband)
+  ) %>%
+  select(
+    chromosome_arm,
+    # Distinguish this dominant status that is based on cytobands, from the status
+    dominant_cytoband_status = dominant_status,
+    cytoband,
+    Kids_First_Biospecimen_ID,
+    band_length,
+    gain_fraction,
+    loss_fraction,
+    callable_fraction
+  )
+```
+
+Read in the gene-level data. 
+
+```{r}
+# Read in the annotated gene level CN file
+consensus_seg_autosomes_df <-
+  read_tsv(file.path(results_dir, "consensus_seg_annotated_cn_autosomes.tsv.gz"))
+```
+
+## Define most focal units
+
+### Determine chromosome arm status
+
+```{r}
+# Use our custom function to make the arm status calls
+arm_status_count <- status_majority_caller(
+  consensus_seg_cytoband_status_df,
+  "chromosome_arm",
+  status_threshold
+)
+
+# Display table
+arm_status_count %>%
+  group_by(dominant_status)
+```
+
+These are the chromosome arms that have not been defined as gain or loss -- we want to define their cytoband/gene-level status
+
+```{r}
+# Let's get a vector of the neutral arms
+neutral_arms <- arm_status_count %>%
+  filter(dominant_status == "neutral") %>%
+  dplyr::pull("chromosome_arm")
+```
+
+### Determine cytoband status
+
+We want to include cytoband and gene-level calls for chromosome arms that have not been defined as a gain or loss to make the cytoband-level majority calls.
+
+```{r}
+# Now count the cytoband level calls (for each status call) and define
+# the cytoband as that status if more than 90% of the total counts are
+# for that particular status
+cytoband_status_count <- consensus_seg_cytoband_status_df %>%
+  # Filter the annotated CN data to include only neutral chromosome arms
+  filter(chromosome_arm %in% neutral_arms) %>%
+  select(
+    Kids_First_Biospecimen_ID,
+    cytoband,
+    loss_fraction,
+    gain_fraction,
+    callable_fraction,
+    dominant_status = dominant_cytoband_status
+  ) %>%
+  mutate(
+    dominant_status = case_when(
+      callable_fraction < (1 - uncallable_threshold) ~ "uncallable",
+      loss_fraction > status_threshold ~ "loss",
+      gain_fraction > status_threshold ~ "gain",
+      loss_fraction + gain_fraction > status_threshold ~ "unstable",
+      TRUE ~ "neutral"
+    )
+  )
+
+# Display table
+cytoband_status_count %>%
+  group_by(dominant_status)
+```
+
+### Determine gene-level status
+
+```{r}
+# These are the cytobands that have not been defined as gain or loss --
+# we want to define their gene-level status
+neutral_cytobands <- cytoband_status_count %>%
+  filter(dominant_status %in% c("unstable", "neutral")) %>%
+  dplyr::pull("cytoband")
```

Comment:
Similarly to above, I think we want to do this via a `left_join` and comparisons within each row. The reason being we need to preserve the relationship between samples and cytobands, not just keep all cytobands with at least one neutral call.

Reply:
Note that when we join these genes to cytobands, some of the cytoband columns in `consensus_seg_cytoband_status_df` have more than one cytoband as genes can cross boundaries. We will need to add logic to fill those in as possible.

We will also have to be careful here to be sure the logic flows correctly, as we don't want to make mistakes in filtering at this point because of any filtering we might have done at the cytoband level! That is, if we filtered a cytoband because it was the same call as the arm that contains it, we still want to filter out the genes within that cytoband that have the same call!